airbyte-bootloader/src/main/java/io/airbyte/bootloader/Bootloader.java

```diff
@@ -0,0 +1,291 @@
+                    final Optional<DefinitionsProvider> definitionsProvider,
+                    final FeatureFlags featureFlags,
+                    @Named("jobsDatabaseInitializer") final DatabaseInitializer jobsDatabaseInitializer,
+                    @Named("jobsDatabaseMigrator") final DatabaseMigrator jobsDatabaseMigrator,
+                    final JobPersistence jobPersistence,
+                    final ProtocolVersionChecker protocolVersionChecker,
+                    @Value("${airbyte.run-migration-on-startup}") final boolean runMigrationOnStartup,
+                    final SecretMigrator secretMigrator) {
+    this(autoUpgradeConnectors, configRepository, configsDatabaseInitializer, configsDatabaseMigrator, currentAirbyteVersion,
+        definitionsProvider, featureFlags, jobsDatabaseInitializer, jobsDatabaseMigrator, jobPersistence, protocolVersionChecker,
+        runMigrationOnStartup, secretMigrator, null);
+  }
+
+  @VisibleForTesting
+  Bootloader(
+             @Value("${airbyte.protocol.auto-upgrade-connectors}") final boolean autoUpgradeConnectors,
+             final ConfigRepository configRepository,
+             @Named("configsDatabaseInitializer") final DatabaseInitializer configsDatabaseInitializer,
+             @Named("configsDatabaseMigrator") final DatabaseMigrator configsDatabaseMigrator,
+             final AirbyteVersion currentAirbyteVersion,
+             final Optional<DefinitionsProvider> definitionsProvider,
+             final FeatureFlags featureFlags,
+             @Named("jobsDatabaseInitializer") final DatabaseInitializer jobsDatabaseInitializer,
+             @Named("jobsDatabaseMigrator") final DatabaseMigrator jobsDatabaseMigrator,
+             final JobPersistence jobPersistence,
+             final ProtocolVersionChecker protocolVersionChecker,
+             @Value("${airbyte.run-migration-on-startup}") final boolean runMigrationOnStartup,
+             final SecretMigrator secretMigrator,
+             final Runnable postLoadExecution) {
+    this.autoUpgradeConnectors = autoUpgradeConnectors;
+    this.configRepository = configRepository;
+    this.configsDatabaseInitializer = configsDatabaseInitializer;
+    this.configsDatabaseMigrator = configsDatabaseMigrator;
+    this.currentAirbyteVersion = currentAirbyteVersion;
+    this.definitionsProvider = definitionsProvider;
+    this.featureFlags = featureFlags;
+    this.jobsDatabaseInitializer = jobsDatabaseInitializer;
+    this.jobsDatabaseMigrator = jobsDatabaseMigrator;
+    this.jobPersistence = jobPersistence;
+    this.protocolVersionChecker = protocolVersionChecker;
+    this.runMigrationOnStartup = runMigrationOnStartup;
+    this.secretMigrator = secretMigrator;
+    this.postLoadExecution = postLoadExecution;
+  }
+
+  /**
+   * Ensures additional initialization is performed after all dependency injection has completed.
+   */
+  @PostConstruct
+  public void afterInitialization() {
+    postLoadExecution = () -> {
```

Comment:
This was initialized in the constructor because cloud had a different one. With the new `ApplyDefinitionsHelper` I think we should be close to have the different behaviors parameterized rather than injected.

Reply:
In DI frameworks, it is not advisable to initialize code that depends on other singletons in a constructor, as those objects may also not be completely initialized yet.  Regardless, I think the plan will be to move this to a concrete class that implements `Runnable` and is its own singleton, so that cloud can just provide a replacement.

Reply:
Yes, different implementation was what I was thinking. I think some recent changes took us halfway already since both behaviors now live in this repo. I guess the next question might be wether we want to keep the flexibility of a post load hook that can do anything.

Reply:
This has now been extracted to an interface/implementation that can be overridden in cloud.